flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java

```diff
@@ -0,0 +1,207 @@
+ */
+
+package org.apache.flink.runtime.state.ttl;
+
+import org.apache.flink.api.common.state.AggregatingStateDescriptor;
+import org.apache.flink.api.common.state.FoldingStateDescriptor;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ReducingStateDescriptor;
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.StateDescriptor;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeutils.CompositeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.KeyedStateFactory;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This state factory wraps state objects, produced by backends, with TTL logic.
+ */
+public class TtlStateFactory {
+	public static <N, SV, S extends State, IS extends S> IS createStateAndWrapWithTtlIfEnabled(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc,
+		KeyedStateFactory originalStateFactory,
+		TtlConfig ttlConfig,
+		TtlTimeProvider timeProvider) throws Exception {
+		return ttlConfig.getTtlUpdateType() == TtlUpdateType.Disabled ?
+			originalStateFactory.createState(namespaceSerializer, stateDesc) :
+			new TtlStateFactory(originalStateFactory, ttlConfig, timeProvider)
+				.createState(namespaceSerializer, stateDesc);
+	}
+
+	private final Map<Class<? extends StateDescriptor>, StateFactory> stateFactories;
+
+	private final KeyedStateFactory originalStateFactory;
+	private final TtlConfig ttlConfig;
+	private final TtlTimeProvider timeProvider;
+
+	private TtlStateFactory(KeyedStateFactory originalStateFactory, TtlConfig ttlConfig, TtlTimeProvider timeProvider) {
+		this.originalStateFactory = originalStateFactory;
+		this.ttlConfig = ttlConfig;
+		this.timeProvider = timeProvider;
+		this.stateFactories = createStateFactories();
+	}
+
+	private Map<Class<? extends StateDescriptor>, StateFactory> createStateFactories() {
+		return Stream.of(
+			Tuple2.of(ValueStateDescriptor.class, (StateFactory) this::createValueState),
+			Tuple2.of(ListStateDescriptor.class, (StateFactory) this::createListState),
+			Tuple2.of(MapStateDescriptor.class, (StateFactory) this::createMapState),
+			Tuple2.of(ReducingStateDescriptor.class, (StateFactory) this::createReducingState),
+			Tuple2.of(AggregatingStateDescriptor.class, (StateFactory) this::createAggregatingState),
+			Tuple2.of(FoldingStateDescriptor.class, (StateFactory) this::createFoldingState)
+		).collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+	}
+
+	private interface StateFactory {
```

Comment:
Do we really need this interface? It looks identical to `KeyedStateFactory` and seems that interface would fit here as well?

Reply:
Leftover from previous iteration 👍